test/select.jl

```diff
@@ -0,0 +1,838 @@
+    @test_throws ArgumentError select(df, :x1, :x5)
+    @test select(df, :x2, r"x", :x1, :) == df[:, [:x2, :x1, :x3, :x4]]
+
+    @test_throws ArgumentError select(df, :x1, :x2 => :x1)
+    @test_throws ArgumentError select(df, :x3 => :x1, :x2 => :x1)
+    @test_throws ArgumentError select(df, :x1, :x2 => identity => :x1)
+    @test_throws ArgumentError select(df, :x1 => :x1, :x2 => identity => :x1)
+    @test_throws ArgumentError select(df, :x3 => identity => :x1, :x2 => identity => :x1)
+    @test select(df, [:x1], :x2 => :x1) == DataFrame(x1 = df.x2)
+
+    @test_throws ArgumentError select!(df, :x1, :x1)
+    @test_throws ArgumentError select!(df, :x1, :x5)
+    @test df == df_ref
+
+    select!(df, :x2, r"x", :x1, :)
+    @test df == df_ref[:, [:x2, :x1, :x3, :x4]]
+end
+
+@testset "SubDataFrame selection" begin
+    df = DataFrame(rand(12, 5))
+    sdf = view(df, 1:10, 1:4)
+    df_ref = copy(sdf)
+
+    @test select(sdf, :x2, :, :x1 => ByRow(x -> x^2) => :r1, :x1 => (x -> x .^ 2) => :r2,
+                 [:x1, :x2] => (+) => :x1, 1:2 => ByRow(/) => :x3, :x1 => :x4) ==
+          select(df_ref, :x2, :, :x1 => ByRow(x -> x^2) => :r1, :x1 => (x -> x .^ 2) => :r2,
+                 [:x1, :x2] => (+) => :x1, 1:2 => ByRow(/) => :x3, :x1 => :x4)
+
+    for fun in (+, ByRow(+))
+        @test select(sdf, 1 => fun) ==
+              DataFrame(Symbol("x1_+") => sdf.x1)
+        @test select(sdf, 1:2 => fun) ==
+              DataFrame(Symbol("x1_x2_+") => sdf.x1 + sdf.x2)
+        @test select(sdf, 1:3 => fun) ==
+              DataFrame(Symbol("x1_x2_x3_+") => sdf.x1 + sdf.x2 + sdf.x3)
+        @test select(sdf, 1:4 => fun) ==
+              DataFrame(Symbol("x1_x2_etc_+") => sum.(eachrow(sdf)))
+    end
+
+    @test_throws ArgumentError select(sdf, :x1, :x1)
+    @test_throws ArgumentError select(sdf, :x1, :x1, copycols=false)
+    @test select(sdf, :x1, [:x1]) == sdf[:, [:x1]]
+    @test select(sdf, :x1, [:x1]) isa DataFrame
+    @test select(sdf, :x1, [:x1], copycols=false) == sdf[:, [:x1]]
+    @test select(sdf, :x1, [:x1], copycols=false) isa SubDataFrame
+    @test_throws ArgumentError select(sdf, :x1 => :r1, copycols=false)
+    @test_throws ArgumentError select(sdf, :x1 => identity => :r1, copycols=false)
+end
+
+@testset "copycols special cases" begin
+    df = DataFrame(a=1:3, b=4:6)
+    c = [7, 8]
+    df2 = select(df, :a => (x -> c) => :c1, :b => (x -> c) => :c2)
+    @test df2.c1 === df2.c2
+    df2 = select(df, :a => identity => :c1, :a => :c2)
+    @test df2.c1 !== df2.c2
+    df2 = select(df, :a => identity => :c1)
+    @test df2.c1 !== df.a
+    df2 = select(df, :a => (x -> df.b) => :c1)
+    @test df2.c1 === df.b
+    df2 = select(view(df, 1:2, :), :a => parent => :c1)
+    @test df2.c1 !== df.a
+    df2 = select(view(df, 1:2, :), :a => (x -> view(x, 1:1)) => :c1)
+    @test df2.c1 isa Vector
+    df2 = select(df, :a, :a => :b, :a => identity => :c, copycols=false)
+    @test df2.b === df2.c == df.a
```

Comment:
Test that they are both `!== a`? Same below.

Reply:
actually here they should be `=== df.a`. I will fix the tests